agent.go

```diff
@@ -0,0 +1,461 @@
+		p.process.Stderr.(*os.File).Close()
+	}
+
+	if p.consoleSock != nil {
+		p.process.Stderr.(*os.File).Close()
+	}
+}
+
+// This is the list of file descriptors we can properly close after the process
+// has exited. These are the remaining file descriptors that we have opened and
+// are no longer needed.
+func (p *process) closePostExitFDs() {
+	if p.termMaster != nil {
+		p.termMaster.Close()
+	}
+
+	if p.stdin != nil {
+		p.stdin.Close()
+	}
+
+	if p.stdout != nil {
+		p.stdout.Close()
+	}
+
+	if p.stderr != nil {
+		p.stderr.Close()
+	}
+}
+
+func (c *container) setProcess(pid string, process *process) {
+	c.Lock()
+	c.processes[pid] = process
+	c.Unlock()
+}
+
+func (c *container) deleteProcess(pid string) {
+	c.Lock()
+	delete(c.processes, pid)
+	c.Unlock()
+}
+
+func (c *container) removeContainer() error {
+	c.wgProcesses.Wait()
+
+	if err := c.container.Destroy(); err != nil {
+		return err
+	}
+
+	return unmountContainerRootFs(c.id, c.config.Rootfs)
+}
+
+func (c *container) getProcess(pid string) (*process, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	proc, exist := c.processes[pid]
+	if !exist {
+		return nil, fmt.Errorf("Process %s not found (container %s)", pid, c.id)
+	}
+
+	return proc, nil
+}
+
+func (p *pod) getContainer(id string) (*container, error) {
+	p.RLock()
+	defer p.RUnlock()
+
+	ctr, exist := p.containers[id]
+	if !exist {
+		return nil, fmt.Errorf("Container %s not found", id)
+	}
+
+	return ctr, nil
+}
+
+func (p *pod) setContainer(id string, ctr *container) {
+	p.Lock()
+	p.containers[id] = ctr
+	p.Unlock()
+}
+
+func (p *pod) deleteContainer(id string) {
+	p.Lock()
+	delete(p.containers, id)
+	p.Unlock()
+}
+
+func (p *pod) getRunningProcess(cid, pid string) (*process, *container, error) {
+	if p.running == false {
+		return nil, nil, fmt.Errorf("Pod not started")
+	}
+
+	ctr, err := p.getContainer(cid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	status, err := ctr.container.Status()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if status != libcontainer.Running {
+		return nil, nil, fmt.Errorf("Container %s %s, should be %s", cid, status.String(), libcontainer.Running.String())
+	}
+
+	proc, err := ctr.getProcess(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return proc, ctr, nil
+}
+
+func (p *pod) readStdio(cid, pid string, length int, stdout bool) ([]byte, error) {
+	proc, _, err := p.getRunningProcess(cid, pid)
+	if err != nil {
+		return nil, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		if stdout {
+			file = proc.stdout
+		} else {
+			file = proc.stderr
+		}
+	}
+
+	buf := make([]byte, length)
+
+	if _, err := file.Read(buf); err != nil {
+		return nil, err
+	}
+
+	return buf, nil
+}
+
+func (p *pod) initLogger() error {
+	agentLog.Logger.Formatter = &logrus.JSONFormatter{TimestampFormat: time.RFC3339Nano}
+
+	config := newConfig(defaultLogLevel)
+	if err := config.getConfig(kernelCmdlineFile); err != nil {
```

Comment:
looks quite strange to get log config from `kernelCmdlineFile` though looks it works correctly.